File: src/main/java/com/clearcont/clearcontapp/views/components/details/BalanceteDetailsLayout.java
package com.clearcont.clearcontapp.views.components.details;

import com.clearcont.clearcontapp.model.Balancete;
import com.clearcont.clearcontapp.model.ComposicaoLancamentosContabeis;
import com.clearcont.clearcontapp.service.AnexoStorageServiceImpl;
import com.vaadin.flow.component.Text;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.H3;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.orderedlayout.FlexComponent;
import com.vaadin.flow.component.orderedlayout.FlexLayout;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.dom.Style;

import static com.clearcont.clearcontapp.helpers.DecimalFormatBR.getDecimalFormat;


public class BalanceteDetailsLayout extends HorizontalLayout {
    FlexLayout diferencaLayout = new FlexLayout();
    Text diferencaValue = new Text("0.00");

    public BalanceteDetailsLayout(Balancete balancete, ComposicaoLancamentosContabeis conciliacao, double saldoContabil, AnexoStorageServiceImpl anexoStorageService) {
        FlexLayout accountName = createStyledLayout("Nome da conta: ", balancete.getNomeConta());
        FlexLayout accountNumber = createStyledLayout("Numero da conta: ", String.valueOf(balancete.getNumeroConta()));
        FlexLayout conciliationStatusLabel = createStyledLayout("Status conciliação: ", conciliacao.getStatus().getName());
        FlexLayout composicaoSaldoLayout = createStyledLayout("Composição do Saldo Contábil: ", getDecimalFormat().format(saldoContabil));

        Span differenceSpan = new Span(getDecimalFormat().format(balancete.getDoubleTotalBalancete() - saldoContabil));
        differenceSpan.getStyle().set("font-weight", "800").set("padding-left", "5px");
        var newC = new HorizontalLayout(new Span("Diferença: "), differenceSpan);
        diferencaLayout.replace(diferencaValue, newC);

        FlexLayout infos = new FlexLayout(
                accountName,
                accountNumber,
                conciliationStatusLabel,
                composicaoSaldoLayout,
                diferencaLayout
        );
        infos.addClassName("card");
        infos.setAlignItems(Alignment.START);
        infos.setFlexDirection(FlexLayout.FlexDirection.COLUMN);

        H3 documentosAnexadosTitle = new H3("Documentos Anexados");
        documentosAnexadosTitle.getStyle().setTextAlign(Style.TextAlign.CENTER);
        Div documentosAnexadosDiv = new Div(
                documentosAnexadosTitle,
                new DownloadComponent(anexoStorageService, conciliacao, balancete.getEmpresa().getNomeEmpresa())
        );
        documentosAnexadosDiv.addClassName("card");
        documentosAnexadosDiv.setMinWidth("200px");
        FlexLayout infosCards = new FlexLayout(documentosAnexadosDiv, infos);
        infosCards.getStyle().setPadding("10px");
        infosCards.setFlexWrap(FlexLayout.FlexWrap.WRAP);
        infosCards.setFlexShrink(20);
        infosCards.setJustifyContentMode(JustifyContentMode.AROUND);
        add(infosCards);
    }

    public void updateDiferencaLayout(double saldoContabilBalancete, double saldoContabilDaGrid) {
        double diferenca = saldoContabilBalancete - saldoContabilDaGrid;
        diferencaValue.setText("R$ " + (getDecimalFormat().format(diferenca)));
    }

    private FlexLayout createStyledLayout(String label, String value) {
        Span span = new Span(value);
        span.getStyle().set("font-weight", "800").set("padding-left", "5px");
        return new FlexLayout(new Span(label), span);
    }
}

File: src/main/kotlin/br/com/clearcont/clearcontwebapp/views/components/details/BalanceteDetailsLayout.kt
package br.com.clearcont.clearcontwebapp.views.components.details

import br.com.clearcont.clearcontwebapp.models.Balancete
import br.com.clearcont.clearcontwebapp.models.ComposicaoLancamentosContabeis
import br.com.clearcont.clearcontwebapp.services.GridConciliarListener
import br.com.clearcont.clearcontwebapp.services.impl.FileUploadServiceImplement
import br.com.clearcont.clearcontwebapp.utils.helpers.formatCurrencyBR
import com.vaadin.flow.component.Composite
import com.vaadin.flow.component.Text
import com.vaadin.flow.component.UI
import com.vaadin.flow.component.html.Div
import com.vaadin.flow.component.html.H3
import com.vaadin.flow.component.html.Span
import com.vaadin.flow.component.orderedlayout.FlexComponent
import com.vaadin.flow.component.orderedlayout.FlexLayout
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.dom.Style
import com.vaadin.flow.spring.annotation.UIScope
import org.slf4j.LoggerFactory


@UIScope
open class BalanceteDetailsLayout(
    private val balancete: Balancete,
    private val conciliacao: ComposicaoLancamentosContabeis,
    private val saldoContabil: Double,
    private val anexoStorageService: FileUploadServiceImplement?,
    private val ui: UI
) : Composite<HorizontalLayout>(), GridConciliarListener {
    private val log = LoggerFactory.getLogger(BalanceteDetailsLayout::class.java)
    private lateinit var diferencaLayout: FlexLayout
    private var diferencaValue = Text("Diferença: ${formatCurrencyBR(saldoContabil)}")

    init {
        buildLayout()
    }

    private fun buildLayout() {
        diferencaLayout = FlexLayout().apply { this.add(diferencaValue) }

        val accountName = createStyledLayout("Nome da conta: ", balancete.nomeConta)
        val accountNumber = createStyledLayout("Numero da conta: ", balancete.numeroConta.toString())
        val conciliationStatusLabel = createStyledLayout("Status conciliação: ", conciliacao.status?.value)
        val composicaoSaldoLayout =
            createStyledLayout("Composição do Saldo Contábil: ", formatCurrencyBR(saldoContabil))

        calculateDifference(saldoContabil)

        val infos = FlexLayout(
            accountName,
            accountNumber,
            conciliationStatusLabel,
            composicaoSaldoLayout,
            diferencaLayout
        ).apply {
            addClassName("card")
            alignItems = FlexComponent.Alignment.START
            flexDirection = FlexLayout.FlexDirection.COLUMN
        }

        val documentosAnexadosTitle = H3("Documentos Anexados").apply {
            style.setTextAlign(Style.TextAlign.CENTER)
        }

        val documentosAnexadosDiv = Div(
            documentosAnexadosTitle,
            anexoStorageService?.let {
                balancete.empresa?.nomeEmpresa?.let { companyName ->
                    DownloadComponent(it, conciliacao, companyName)
                }
            }
        ).apply {
            addClassName("card")
            minWidth = "200px"
        }

        val infosCards = FlexLayout(documentosAnexadosDiv, infos).apply {
            style.setPadding("10px")
                .setBackgroundColor("azure")
                .setBorder("1px solid black")
                .setBorderRadius("30px")
            flexWrap = FlexLayout.FlexWrap.WRAP
            setFlexShrink(20.0)
            justifyContentMode = FlexComponent.JustifyContentMode.AROUND
        }

        content.add(infosCards)
    }

    private fun calculateDifference(saldoContabil: Double) {
        diferencaValue.text = "Diferença: ${formatCurrencyBR(saldoContabil)}"
    }

    private fun refreshLayout() {
        log.info("Refreshing layout")
        content.removeAll()
        buildLayout()
        ui.push()
        log.info("UI pushed")
    }

    private fun createStyledLayout(label: String, value: String?): FlexLayout {
        val span = Span(value).apply {
            style.set("font-weight", "800")["padding-left"] = "5px"
        }
        return FlexLayout(Span(label), span)
    }

    override fun onSaldoContabilChanged(saldoContabil: Double) {
        ui.access {
            diferencaValue.text = "Diferença: ${formatCurrencyBR(saldoContabil)}"
            ui.push()
            refreshLayout()
            log.info("Atualizando detalhes: Saldo Contábil Balancete: $saldoContabil")
        }
    }

}
